nemo/utils/callbacks/nemo_model_checkpoint.py

```diff
@@ -271,14 +393,23 @@ def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str)
                 super()._save_checkpoint(trainer, filepath)
         else:
             super()._save_checkpoint(trainer, filepath)
+        # barrier_before=True, so all ranks synchronize before removing the unfinished checkpoint marker
+        # we don't want to remove the marker until all checkpointing is done.
+        self.remove_checkpoint_unfinished_marker(filepath, barrier_before=True)
 
     def _remove_checkpoint(self, trainer: "pytorch_lightning.Trainer", filepath: str) -> None:
+        # barrier_after=True, so all ranks continue after the unfinished checkpoint marker is placed.
+        # if anything goes wrong during removal, we should be able to detect that data is incomplete.
+        self.set_checkpoint_unfinished_marker(filepath, barrier_after=True)
         super()._remove_checkpoint(trainer, filepath)
         ema_callback = self._ema_callback(trainer)
         if ema_callback is not None:
             # remove EMA copy of the state dict as well.
             filepath = self._ema_format_filepath(filepath)
             super()._remove_checkpoint(trainer, filepath)
+        # barrier_before=True, so all ranks synchronize before removing the unfinished checkpoint marker
+        # we don't want to remove the marker until the checkpoint is actually removed.
+        self.remove_checkpoint_unfinished_marker(filepath, barrier_before=True)
```

Comment:
@jbieniusiewi do we want to have `barrier_after` in `remove_checkpoint_unfinished_marker` ? Similarly `barrier_before` in case of `set_checkpoint_unfinished_marker`. I am not able to understand the use case of it, since we always want to have `barrier_after=true` for `set_checkpoint_unfinished_marker` and `barrier_before=true` for `remove_checkpoint_unfinished_marker`.

Reply:
Hi @athitten 
Thank you for the review!

Please let me clarify the purpose of the barriers, with `_save_checkpoint`
```
 def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str) -> None:
        self.set_checkpoint_unfinished_marker(filepath, barrier_after=True)

        # (...) checkpoint writing code (...)
        
        self.remove_checkpoint_unfinished_marker(filepath, barrier_before=True)
```
The goal of `self.set_checkpoint_unfinished_marker(filepath, **barrier_after=True**)`
is to ensure that when ranks start to write the checkpoint, the [checkpoint] unfinished marker file is already put in the filesystem. In other words, any checkpointing writing starts **after** the marker is placed.
Without the barrier, some ranks might write some data, before the marker file is placed [by rank0]; in such case, if something crashed before the marker was actually placed, we would get incomplete data and no [unfinished] marker.

On the other hand `self.remove_checkpoint_unfinished_marker(filepath, **barrier_before=True**)`
ensures that all ranks finishes the checkpoint writing **before** we remove the unfinished marker file. Without the barrier, it would be possible that if checkpointing crashed after the marker was removed,  we would get incomplete data and no marker.

I hope this helps!